### PR TITLE
zephyr/i2c: Add support for hardware i2c

### PR DESCRIPTION
**Pull request by MaureenHelm**
Adds support for hardware i2c to the zephyr port. Similar to other ports
such as stm32 and nrf, we only implement the i2c protocol functions
(readfrom and writeto) and defer memory operations (readfrom_mem,
readfrom_mem_into, and writeto_mem) to the software i2c implementation.
This may need to change in the future because zephyr is considering
deprecating its i2c_transfer function in favor of i2c_write_read; in
this case we would probably want to implement the memory operations
directly using i2c_write_read.

Tested with the accelerometer on frdm_k64f and bbc_microbit boards.

### CODE REVIEW

**Comment by MaureenHelm** [comment]
cc: @pfalcon

File: ports/zephyr/machine_i2c.c [hunk 1]
```diff
@@ -0,0 +1,145 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2013, 2014, 2015 Damien P. George
+ * Copyright (c) 2019, NXP
```
Comment (pfalcon) [inline ports/zephyr/machine_i2c.c]: Superfluous comma?

File: ports/zephyr/machine_i2c.c [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        { MP_QSTR_timeout, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_obj = MP_OBJ_NULL} },
+    };
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    const char *dev_name = mp_obj_str_get_str(args[ARG_id].u_obj);
+    struct device *dev = device_get_binding(dev_name);
+
+    if (!dev) {
```
Comment (pfalcon) [inline ports/zephyr/machine_i2c.c]: MicroPython codestyle is the same what's now requested (but still regularly ignored) for Zephyr: to use explicit comparison `dev == NULL`.

File: ports/zephyr/machine_i2c.c [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+STATIC int machine_hard_i2c_transfer(mp_obj_base_t *self_in, uint16_t addr, uint8_t *buf, size_t len, bool stop, bool read) {
+    machine_hard_i2c_obj_t *self = (machine_hard_i2c_obj_t *)self_in;
+    struct i2c_msg msg;
+    int ret;
+
+    msg.buf = (u8_t *)buf;
```
Comment (pfalcon) [inline ports/zephyr/machine_i2c.c]: Unlike Zephyr, MicroPython codestyle avoids that strange-looking space in a cast, i.e. it should be `(u8_t*)`.

**Comment by pfalcon** [comment]
@stinos, I believe appveyor CI support was added by you, right? Can you please look what's wrong with it here - this patch doesn't touch anything outside Zephyr port, while appveyor fails, seemingly producing zero log output. I suggested to force-push to retrigger the build, and appveyor failed in the same way as before.

@MaureenHelm: Perhaps, try to repush again now, maybe that was a bad day for appveyor? Recently posted PRs doesn't show failures.

**Comment by stinos** [comment]
Seems there was an Appveyor outage about 3 days ago (see https://help.appveyor.com/discussions/problems/22315-builds-are-failing-most-of-the-time-fallback-clouds-stack-is-too-big), should be up again now.

File: ports/zephyr/machine_i2c.c (hunk 1 above)
Comment (MaureenHelm) [reply ports/zephyr/machine_i2c.c]: No

**Comment by MaureenHelm** [comment]
@pfalcon Thanks for the review. I pushed a new version

**Comment by dpgeorge** [comment]
Thanks very much @MaureenHelm , a good addition!  And thanks @pfalcon for the review.  Merged in 2befcb8a9d54f3f9c059b45eba3e9d4e5fb13514

File: ports/zephyr/prj_base.conf [hunk 4]
```diff
@@ -14,6 +14,9 @@ CONFIG_NEWLIB_LIBC=y
 CONFIG_FLOAT=y
 CONFIG_MAIN_STACK_SIZE=4736
 
+# Drivers
+CONFIG_I2C=y
```
Comment (pfalcon) [inline ports/zephyr/prj_base.conf]: Postreview: this is actually not a good change. Support for I2C (or not) is still a board-specific matter. This change broke CI build for qemu_cortex_m3, which doesn't have I2C support. (Or course, we might argue that Zephyr shouldn't error the build in this case, but we have what we have anyway.)

**Review by pfalcon — COMMENTED** [review]
@MaureenHelm, Thanks for this patch. Sorry, I only now get to look into it at more detail and test. There're also a couple of problem spots (I'm going to fix them in my repo).

From commit message:

> This may need to change in the future because zephyr is considering
deprecating its i2c_transfer function in favor of i2c_write_read; in
this case we would probably want to implement the memory operations
directly using i2c_write_read.

I hope you would use the experience in developing this module to try to protect i2c_transfer() availability in Zephyr TSC meetings ;-). There're really too much shufflings and deprecations happening in Zephyr. If it works and offers more flexibility, why deprecate it?

Otherwise, I tested this class on FRDM-K64F with its FXOS8700 sensor (I actually posted on this yesterday, but don't see my comment today, weird).

Here're some commands from my testing session, for future reference:

~~~
from machine import I2C
i2c = I2C("I2C_0")
# ID reg
i2c.writeto(29, b"\x0d", False)
i2c.readfrom(29, 1)
b'\xc7'

i2c.readfrom_mem(29, 0x0d, 1)
b'\xc7'

# Magnetometer readings
ubinascii.hexlify(i2c.readfrom_mem(29, 0x33, 6), " ")
~~~

File: ports/zephyr/README.md [hunk 5]
```diff
@@ -81,6 +82,13 @@ To blink an LED:
         LED.value(0)
         time.sleep(0.5)
 
+To scan for I2C slaves:
```
[the remark this answers is not in the document]
Comment (MaureenHelm) [reply ports/zephyr/README.md]: Where should it go instead?

File: ports/zephyr/prj_base.conf (hunk 4 above)
Comment (MaureenHelm) [reply ports/zephyr/prj_base.conf]: Should support for the I2C micropython module be conditional then?

**Comment by MaureenHelm** [comment]
> > This may need to change in the future because zephyr is considering
> > deprecating its i2c_transfer function in favor of i2c_write_read; in
> > this case we would probably want to implement the memory operations
> > directly using i2c_write_read.
> 
> I hope you would use the experience in developing this module to try to protect i2c_transfer() availability in Zephyr TSC meetings ;-). There're really too much shufflings and deprecations happening in Zephyr. If it works and offers more flexibility, why deprecate it?

`i2c_transfer()` isn't well defined and behaves differently on different platforms, particularly around restart conditions. I considered not using it at all in favor of `i2c_write_read()` instead, but that would make zephyr the only hard i2c implementation in micropython that doesn't defer to soft i2c for some functions. I wasn't sure if that would be acceptable, so I went with the simpler approach for my first PR. I can send another PR if it is.

File: ports/zephyr/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (pfalcon) [reply ports/zephyr/README.md]: > Where should it go instead?

In https://github.com/micropython/micropython/pull/4668, I've reordered it so it go after the text related to GPIO.

File: ports/zephyr/prj_base.conf (hunk 4 above)
Comment (pfalcon) [reply ports/zephyr/prj_base.conf]: > Should support for the I2C micropython module be conditional then?

I'm not sure, just switching back to enable it per-board fixed the build, I didn't look further so far. That's what included in https://github.com/micropython/micropython/pull/4668.

**Comment by dpgeorge** [comment]
> I considered not using it at all in favor of `i2c_write_read()` instead, but that would make zephyr the only hard i2c implementation in micropython that doesn't defer to soft i2c for some functions

What's important is that a port does it's best to implement the defined I2C methods.  Whether it does that by providing readfrom/writeto C-level functions and using the existing Python wrappers, or by implementing everything from scratch, is less important (except that it's much simpler to implement readfrom/witeto C functions as was done).

**Pull request closed without merging**